### PR TITLE
drivers: gpio: test device

### PR DESCRIPTION
**Pull request by JordanYates**
Implements a dummy GPIO driver for the vnd,gpio compatible to allow drivers that use GPIO to be compiled for generic platforms in build_all tests.

This driver follows the conventions of spi_test.c and i2c_test.c, and allows the manually device creation in each "build_all" test to be removed.

### CODE REVIEW

**Comment by JordanYates** [comment]
@nordic-krch, let me know if I've misunderstood the problem with `log_strdup`.

`tests/drivers/build_all/modem` is compiling in `subsys/net/ip/route.c` which uses `log_strdup` without `CONFIG_LOG`, leading to undefined references to `z_log_strdup`.

**Comment by nordic-krch** [comment]
Is this `log_strdup` call happens outside of log macro? I never seen issue like that when it is used inside macro (likg `LOG_INF`) because macro has static if case which prevents linker from attempting to link `log_strdup`. Is it something new?

**Comment by JordanYates** [comment]
> Is this `log_strdup` call happens outside of log macro? I never seen issue like that when it is used inside macro (likg `LOG_INF`) because macro has static if case which prevents linker from attempting to link `log_strdup`. Is it something new?

It is inside of a `NET_DBG`, not sure if that will cause problems. Its been that way for 3 years apparently.

https://github.com/zephyrproject-rtos/zephyr/blob/e1b845cc8875a0f69f2e0a1e7670327a78a1cbd3/subsys/net/ip/route.c#L429-L436

**Comment by JordanYates** [comment]
So the problem can be reproduced on `main` with:
```
./zephyr/scripts/twister -p intel_adsp_cavs15 -s tests/drivers/build_all/modem/drivers.modem.quectel_bg9x.build
```

**Review by mnkp — APPROVED** [review]

File: drivers/gpio/gpio_test.c [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+static int vnd_gpio_manage_callback(const struct device *port,
+				    struct gpio_callback *cb,
+				    bool set)
+{
+	return -ENOTSUP;
+}
+
+uint32_t vnd_gpio_get_pending_int(const struct device *dev)
+{
+	return -ENOTSUP;
```
Comment (nordic-krch) [inline drivers/gpio/gpio_test.c:72]: should return 0.

Comment (JordanYates) [reply drivers/gpio/gpio_test.c]: Fixed. It's interesting that something so obviously wrong didn't result in compiler warnings/errors.

**Review by nordic-krch — APPROVED** [review]

**Review by mbolivar-nordic — COMMENTED** [review]
Late to the party, but LGTM

**Pull request merged**